docs/DafnyRef/Topics.md

```diff
@@ -34,6 +34,20 @@ const m := set x | n in {-1, 0, 1, 2} && f(n) >= 1
 const m := set x | n in {-1, 0, 1, 2} && g(n) >= 1
 ```
 
+A special case to be aware of: if the type in the comprehension differs from the collection,
+but both are compilable, the type of the comprehension will be fully tested, which could be viewed as a performance problem if only one specific trait needed to be tested. For example:
+
+```dafny
+const s: set<nat> := {1, 2, 3}
+type NonNegative = x | x > 0 witness 1
```

Comment:
What if we change this constraint into `x >= 0`, so it's the same as the `nat` constraint.

Could we then write
```
const m := set t: nat, x: NonNegative | t in s && x == t :: x
```

To avoid any runtime type test?

Reply:
> To avoid any runtime type test?

Yes, that would avoid any runtime test because tests are used only for filtering, in your case, the verifier would check that the two constraints are equivalent.